spring-integration-core/src/main/java/org/springframework/integration/mapping/AbstractHeaderMapper.java

```diff
@@ -151,7 +152,22 @@ else if (NON_STANDARD_HEADER_NAME_PATTERN.equals(pattern)) {
 				matchers.add(new PrefixBasedMatcher(false, this.standardHeaderPrefix));
 			}
 			else {
-				matchers.add(new PatternBasedHeaderMatcher(Collections.singleton(pattern)));
+				String thePattern = pattern;
+				boolean negate = false;
+				if (pattern.startsWith("!")) {
+					thePattern = pattern.substring(1);
+					negate = true;
+				}
+				else if (pattern.startsWith("\\!")) {
+					thePattern = pattern.substring(1);
+				}
+				if (negate) {
+					// negative matchers get priority
+					matchers.add(0, new PatternBasedHeaderMatcher(Collections.singleton(thePattern), negate));
+				}
+				else {
+					matchers.add(new PatternBasedHeaderMatcher(Collections.singleton(thePattern), negate));
```

Comment:
Looks like we always use `PatternBasedHeaderMatcher` with the **single** pattern. Therefore we have some overhead with the extra `Collection` object and iteration within `matchHeader()`.
I agree that it wouldn't be so good to break `PatternBasedHeaderMatcher` because really some `AbstractHeaderMapper` may use it as with the `Collection`.
But from other side how about to rework our logic somehow to have only one `PatternBasedHeaderMatcher` with the collection of all those patterns including negate?